File: Communicate_Mac/Communicate/Common/DataInfo.m
//
//  DataHeader.m
//  Communicate
//
//  Created by Hugh Bellamy on 25/05/2015.
//  Copyright (c) 2015 Hugh Bellamy. All rights reserved.
//

#import "DataInfo.h"

#import "DataContent.h"
#import "DataHeaderFooter.h"
#import "DataHeader.h"
#import "DataFooter.h"
#import "ByteHeader.h"

@interface DataInfo ()

@property (assign, readwrite, nonatomic) CommunicationDataType dataType;
@property (assign, readwrite, nonatomic) NSUInteger headerLength;
@property (assign, readwrite, nonatomic) NSUInteger contentLength;
@property (assign, readwrite, nonatomic) NSUInteger footerLength;

+ (NSArray *)byteHeaders;

@end

@implementation DataInfo

- (instancetype)initWithDataType:(CommunicationDataType)dataType header:(DataHeader *)header content:(DataContent *)content footer:(DataFooter *)footer {
    self = [super init];
    if(self) {
        self.dataType = dataType;
        self.headerLength = [header getData].length;
        self.contentLength = [content getData].length;
        self.footerLength = [footer getData].length;
    }
    return self;
}

- (instancetype)initWithData:(NSData *)data {
    self = [super init];
    if(self) {
        self.dataType = [DataInfo dataTypeFromHeaderData:data];
        self.contentLength = [DataInfo contentLengthFromHeaderData:data];
        self.footerLength = [DataInfo footerLengthFromHeaderData:data];
    }
    return self;
}

- (NSData *)getData {
    NSMutableData *header = [[NSMutableData alloc]init];
    
    NSUInteger headerLength = self.headerLength;
    NSUInteger contentLength = self.contentLength;
    NSUInteger footerLength = self.footerLength;
    
    NSData *typeData = [DataInfo typeDataFromDataType:self.dataType];
    NSData *headerLengthData = [NSData dataWithBytes: &headerLength length:4];
    NSData *contentLengthData = [NSData dataWithBytes: &contentLength length:4];
    NSData *footerLengthData = [NSData dataWithBytes: &footerLength length:4];
    
    [header appendData:typeData];
    [header appendData:headerLengthData];
    [header appendData:contentLengthData];
    [header appendData:footerLengthData];
    
    return [NSData dataWithData:header];
}

+ (NSUInteger)length {
    return 14;
}

+ (NSUInteger)headerLengthFromHeaderData:(NSData *)headerData {
    int bytes;
    [headerData getBytes:&bytes range:NSMakeRange(2, 4)];
    return bytes;
}

+ (NSUInteger)contentLengthFromHeaderData:(NSData *)headerData {
    int bytes;
    [headerData getBytes:&bytes range:NSMakeRange(6, 4)];
    return bytes;
}

+ (NSUInteger)footerLengthFromHeaderData:(NSData *)headerData {
    int bytes;
    [headerData getBytes:&bytes range:NSMakeRange(10, 4)];
    return bytes;
}

+ (CommunicationDataType)dataTypeFromHeaderData:(NSData *)headerData {
    char bytes[2];
    [headerData getBytes:&bytes length:sizeof(bytes)];
    
    ByteHeader *byteHeader = [[ByteHeader alloc]initWithFirstByte:bytes[0] secondByte:bytes[1] dataType:CommunicationDataTypeOther];
    ByteHeader *actualByteHeader = [ByteHeader otherByteHeader];
    
    for (ByteHeader *aByteHeader in [DataInfo byteHeaders]) {
        if([aByteHeader isEqual:byteHeader]) {
            actualByteHeader = aByteHeader;
            break;
        }
    }
    return actualByteHeader.dataType;
}


+ (NSData *)typeDataFromDataType:(CommunicationDataType)dataType {
    ByteHeader *byteHeader = [ByteHeader otherByteHeader];
    for(ByteHeader *dataTypeByteHeader in [DataInfo byteHeaders]) {
        if(dataTypeByteHeader.dataType == dataType) {
            byteHeader = dataTypeByteHeader;
            break;
        }
    }
    return [byteHeader toData];
}

+ (NSArray *)byteHeaders {
    static NSArray *byteHeaders;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        byteHeaders = @[[ByteHeader stringByteHeader], [ByteHeader imageByteHeader], [ByteHeader fileByteHeader], [ByteHeader JSONByteHeader], [ByteHeader otherByteHeader]];
    });
    return byteHeaders;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"Data Info: data type = %ld; header length = %lu; content length = %lu; footer length = %lu", (long)self.dataType, (unsigned long)self.headerLength, (unsigned long)self.contentLength, (unsigned long)self.footerLength];
}

@end

File: src/windows/Communicate Core/Data/DataInfo.cs
using System;
using System.Linq;

namespace Communicate
{
    public class DataInfo
    {
        public DataInfo(DataType dataType, DataHeaderFooter header, byte[] content, DataHeaderFooter footer) : 
            this(dataType, 
                Platform.Windows,
                VersionInfo.CurrentVersion,
                GetLength(header?.GetData()), 
                GetLength(content),
                GetLength(footer?.GetData()))
        {
        }

        internal DataInfo(byte[] bytes) :
            this(new DataType(BitConverter.ToInt32(bytes, 0)),
                new Platform(BitConverter.ToInt32(bytes, 4)),
                new VersionInfo(BitConverter.ToInt32(bytes, 8)),
                BitConverter.ToInt32(bytes, 12),
                BitConverter.ToInt32(bytes, 16),
                BitConverter.ToInt32(bytes, 20))
        {
        }

        private DataInfo(DataType dataType, Platform platform, VersionInfo version, int headerLength, int contentLength, int footerLength)
        {
            Type = dataType;
            Platform = platform;
            Version = version;
            HeaderLength = headerLength;
            ContentLength = contentLength;
            FooterLength = footerLength;
        }

        public DataType Type { get; }

        public Platform Platform { get; }
        public VersionInfo Version { get; }

        public int HeaderLength { get; }
        public int ContentLength { get; }
        public int FooterLength { get; }

        public static int DataInfoSize { get; } = 24;

        private static int GetLength(byte[] data) => data?.Length ?? 0;

        public byte[] GetData() =>
            Combine(Type.GetBytes(), 
                    Platform.GetBytes(),
                    Version.GetBytes(),
                    BitConverter.GetBytes(HeaderLength),
                    BitConverter.GetBytes(ContentLength),
                    BitConverter.GetBytes(FooterLength));

        private static byte[] Combine(params byte[][] byteArrays)
        {
            var combined = new byte[byteArrays.Sum(a => a.Length)];
            var offset = 0;
            foreach (var array in byteArrays)
            {
                Buffer.BlockCopy(array, 0, combined, offset, array.Length);
                offset += array.Length;
            }
            return combined;
        }
    }
}